### PR TITLE
add utp to multiaddr-net

### PR DESCRIPTION
**Pull request by whyrusleeping**
it does things and uses utp to do them.

One errata i had to work around was that the `net.Addr`s returned from the utp lib specified 'udp' as their network, so i wrapped them in utp_utils.go and made it return "utp" as the network. Other than that (and me having to remove the go1.5 'for range' syntax) this was a fairly clean integration.

### CODE REVIEW

File: utp_util.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package manet
```
Comment (jbenet) [inline utp_util.go]: this should be a package on its own (can be in this same repo), and keep things like:

```
utp.Addr
utp.Conn
utp.Dialer
utp.Listener
```
(i wouldn't do `utp.UTPConn` anymore, annoying sideeffect of `net` pkg bunching up everything)

Comment (whyrusleeping) [inline utp_util.go]: alright, sounds good to me

File: convert.go [hunk 2]
```diff
@@ -117,7 +117,12 @@ func ToNetAddr(maddr ma.Multiaddr) (net.Addr, error) {
 	case "udp", "udp4", "udp6":
 		return net.ResolveUDPAddr(network, host)
 	case "utp", "utp4", "utp6":
-		return utp.ResolveAddr(network, host)
+		a, err := net.ResolveUDPAddr("udp"+network[3:], host)
+		if err != nil {
+			return nil, err
+		}
+
+		return utp.MakeAddr(a), nil
```
Comment (jbenet) [inline convert.go]: i think it makes sense to have this as `utp.ResolveAddr` as before-- there's a sort of "standard interface" for a `net.Conn` protocol and has all these pieces.

File: net.go [hunk 3]
```diff
@@ -109,24 +110,17 @@ func (d *Dialer) Dial(remote ma.Multiaddr) (Conn, error) {
...
-		// nconn, err = utpd.Dial(rnet, rnaddr)
-		// if err != nil {
-		// 	return nil, err
-		// }
+	case "utp", "utp4", "utp6":
+		// construct utp dialer, with options on our net.Dialer
+		nconn, err = utp.DialTimeout(rnaddr, d.Timeout)
+		if err != nil {
+			return nil, err
+		}
```
Comment (jbenet) [inline net.go]: I think this wants a `Dialer` too, like the other `utp` package. the interface is on the left, commented out:

```go
utpd := utp.Dialer{
  Timeout: d.Dialer.Timeout,
  LocalAddr: d.Dialer.LocalAddr,
}

nconn, err = utpd.Dial(rnet, rnaddr)
if err != nil {
  return nil, err
}
```
it makes it much easier to map other options (like SetDeadline, and so on) if all packages follow a similar set of interfaces as the `net` package. i know anacrolix/utp doesn't, but we can map.

File: net.go [hunk 4]
```diff
@@ -229,9 +223,11 @@ func Listen(laddr ma.Multiaddr) (Listener, error) {
 
 	var nl net.Listener
 	switch lnet {
-	case "utp":
-		// nl, err = utp.Listen(lnet, lnaddr)
-		return nil, fmt.Errorf("utp is currently broken")
+	case "utp", "utp4", "utp6":
+		s, nserr := utp.NewSocket("udp"+lnet[3:], lnaddr)
+		err = nserr
+		nl = &mautp.Listener{s}
```
Comment (jbenet) [inline net.go]: Put thins in a `Listen` call in our wrapper package. multiaddr-net shouldn't have to know all this anacrolix/utp specific stuff.

Comment (jbenet) [inline net.go]: @whyrusleeping here you go:
```go
func Listen(net, addr) (Listener, error) {
  switch {
  case "utp", "utp4", "utp6": 
    s, err := utp.Socket("udp"+net[3:], addr)
    if err != nil {
      return nil, err
    }
    return &mautp.Listener{s}
  default:
    return nil, wrong network err
  }
}
```

**Comment by jbenet** [comment]
LGTM! :+1:  :clap:

**Comment by jbenet** [comment]
will merge wen travis ends

**Pull request merged**